toolchain/parser/parser2.h

```diff
@@ -38,21 +42,49 @@ class Parser2 {
   struct StateStackEntry {
     StateStackEntry(ParserState state, TokenizedBuffer::Token token,
                     int32_t subtree_start)
-        : state(state), token(token), subtree_start(subtree_start) {}
+        : StateStackEntry(state, PrecedenceGroup::ForTopLevelExpression(),
+                          PrecedenceGroup::ForTopLevelExpression(), token,
+                          subtree_start) {}
+
+    StateStackEntry(ParserState state, PrecedenceGroup ambient_precedence,
+                    PrecedenceGroup lhs_precedence,
+                    TokenizedBuffer::Token token, int32_t subtree_start)
+        : state(state),
+          ambient_precedence(ambient_precedence),
+          lhs_precedence(lhs_precedence),
+          token(token),
+          subtree_start(subtree_start) {}
 
     // The state.
     ParserState state;
+    // Set to true  to indicate that an error was found, and that contextual
+    // error recovery may be needed.
+    bool has_error = false;
+
+    // Precedence information used for operator states.
```

Comment:
Maybe expand this a bit to help document how these are used?

Reply:
Added some comments, trying to guess from code. It's not deeply commented in the current Parser.